core/commonMain/src/kotlinx/io/Input.kt

```diff
@@ -340,93 +371,178 @@ public abstract class Input : Closeable {
     }
 
     /**
      * Reads [size] unsigned bytes from an Input and calls [consumer] on each of them.
      * @throws EOFException if no more bytes available.
      */
-    private inline fun readBytes(size: Int, consumer: (unsignedByte: Int) -> Unit) {
+    private fun readBytes(size: Int, consumer: (unsignedByte: Int) -> Unit) {
         var remaining = size
+
+        var current = position
+        var currentLimit = limit
         while (remaining > 0) {
-            if (position == limit) {
-                if (fetchBuffer() == 0) {
+            if (current == currentLimit) {
+                if (fetchCachedOrFill() == 0) {
                     throw EOFException("End of file while reading buffer")
                 }
+
+                current = position
+                currentLimit = limit
             }
-            consumer(buffer.loadByteAt(position++).toInt() and 0xFF)
+
+            val byte = buffer.loadByteAt(current).toInt() and 0xFF
+            consumer(byte)
+
+            current++
             remaining--
         }
+
+        position = current
+        limit = currentLimit
     }
 
     /**
-     * Prepares this Input for reading from the next buffer, either by filling it from the underlying source
-     * or loading from a [previewBytes] after a [preview] operation or if reading from pre-supplied [Bytes]
+     * Allows direct read from a buffer, operates on startOffset + endOffset (exclusive), returns new position.
+     * NOTE: Dangerous to use, if non-local return then position will not be updated.
      *
-     * Current [buffer] should be exhausted at this moment, i.e. [position] should be equal to [limit]
+     * @return consumed bytes count
      */
-    private fun fetchBuffer(): Int {
-        // TODO properly clarify exception message
-        check(position == limit) {
-            // trying to fetch a buffer when previous buffer was not exhausted is an internal error
-            "Throwing bytes away."
+    internal fun readBufferRange(reader: (Buffer, startOffset: Int, endOffset: Int) -> Int): Int {
+        if (position == limit && fetchCachedOrFill() == 0) {
+            return 0
         }
 
-        val discard = previewDiscard
-        val bytes = previewBytes ?: if (discard) {
-            // fast path, no preview operation, reuse current buffer for new data
-            val fetched = fill(buffer)
-            limit = fetched
-            position = 0
-            return fetched
-        } else {
-            // we are collecting bytes, so need to put current buffer to maintain invariant
-            Bytes(bufferPool).also {
-                previewBytes = it
-                it.append(buffer, limit)
-            }
+        val newPosition = reader(buffer, position, limit)
+        val result = newPosition - position
+        position = newPosition
+        return result
+    }
+
+    /**
+     * Copy buffered bytes to [destination]. If no bytes buffered it will block until [fill] is finished.
```

Comment:
`Copies currently buffered bytes to the given [destination]. If no bytes are currently buffered, [fill] method is invoked first`